### DISCUSSION TITLE
NODE_IS_RE_PARENT and NODE_IS_RE_CHILD

### DISCUSSION THREAD

**Discussion by hyanwong** [tskit-dev/what-is-an-arg-paper#275] [Ideas]
I wonder whether some confusion could be avoided if the recombination nodes in a tsARG were labelled either

`tskit.NODE_IS_RE_PARENT` - node represents a genome immediately prior in time to a recombination event (so these come in pairs: this should be an alias for `msprime.NODE_IS_RE_EVENT`)

or 

`tskit.NODE_IS_RE_CHILD` - node represents a genome immediately after a recombination event - we would use this for most of the ARG paper figures

**Comment by JereKoskela** [comment]
I guess we could do that. It isn't clear to me whether more labels will decrease or increase confusion though. As an alternative, we could just draw the two recombination parent nodes on top of each other in figures to achieve the same effect.

**Comment by hyanwong** [comment]
Actually, I wonder if we should have `NODE_IS_RIGHT_RE_PARENT` and `NODE_IS_LEFT_RE_PARENT`, which would help if we needed to match them up.

Re labelling, this isn't so much for labelling plots, more for when we want to convert from a one-node to a two-node version, so we don't get confused about what the nodes represent.

**Reply by hyanwong** [reply]
For example, we could calculate the number of recombination events in an ARG, without having to know how we encoded the recombination nodes, by 

```python
len(np.where(ts.nodes_flags & (NODE_IS_RIGHT_RE_PARENT | NODE_IS_LEFT_RE_PARENT))) / 2
+ len(np.where(ts.nodes_flags & tskit.NODE_IS_RE_CHILD)

**Comment by JereKoskela** [comment]
It isn't clear to me what a one-node version would be for. Presumably the two node approach one can do everything the one node version can, so the question is whether having a one node version adds more clarity than is lost by maintaining two parallel systems for essentially the same thing. The answer doesn't seem obvious to me, and I don't have a strong feeling on what it should be.

**Reply by hyanwong** [reply]
Yes, I guess that's true. However, some people are probably going to want to convert their internal ARG format to a tskit 1-RE-node ARG (I think?), and the 2 node version seems very specifically tailored to the Hudson model. We also want a version that encodes an ARG (i.e. with recombination nodes) produced e.g. by simplifying a large covid tree, or when there is gene conversion. Does it make sense to keep the 2-node encoding for this too? I'm not sure.

**Reply by JereKoskela** [reply]
I don't think the two node version is all that specific to Hudson, in that it can handle multiple recombinations. Gene conversion isn't a problem either---just map some segments to the left parent, then some to the right, and then some left again. It does assume that we are only tracking a single chromosome on the child so that only the two copies on one parent will ever be relevant. I guess for a completely generic picture we'd need 4 parent nodes for a biparental, diploid species.

I'm not sure what the intricacies of Covid trees are, but if they need a one node version, then that does sound like a good reason to do it.

**Comment by jeromekelleher** [comment]
Do we actually need these flags? I.e., can't these nodes be identified by the number of parents and children they have in any case?

**Reply by hyanwong** [reply]
Hmm, good point. The `NODE_IS_RE_CHILD` node can be identified in an full arg because it has multiple parents. The `NODE_IS_RE_PARENT` (or `NODE_IS_RE_LEFT_PARENT`/`NODE_IS_RE_RIGHT_PARENT`) and be identified as the parent(s) of a node with multiple parents.

*However*, this is only true in a "full ARG" simulation, when we are assured either that the `NODE_IS_RE_CHILD` node or alternatively the `NODE_IS_RE_PARENT` node marks the precise time of the recombination event. So the flag is essentially saying "this is a simulation representing a full ARG, and you can treat this node as contemporaneous in time to a recombination event (e.g. for the purposes of likelihood calculation)"

We could flag the entire simulation as a "full arg", but then we would need to remove that flag on simplification, so it seems easier to flag the nodes in the right way and let simplification remove nodes and therefore their flags as necessary.
